pad_server/src/telemetry.c

```diff
@@ -245,26 +246,110 @@ void *telemetry_run(void *arg) {
         header_p hdr = {.type = TYPE_TELEM, .subtype = TELEM_PRESSURE};
         pressure_p body = {.id = 1, .time = time, .pressure = pressure};
 
-        /* Send data to all clients. */
         struct iovec pkt[2] = {
             {.iov_base = &hdr, .iov_len = sizeof(hdr)},
             {.iov_base = &body, .iov_len = sizeof(body)},
         };
         struct msghdr msg = {
-            .msg_name = NULL,
-            .msg_namelen = 0,
             .msg_iov = pkt,
             .msg_iovlen = (sizeof(pkt) / sizeof(struct iovec)),
-            .msg_control = NULL,
-            .msg_controllen = 0,
-            .msg_flags = 0,
         };
         telemetry_publish(&telem, &msg);
-
-        usleep(1000);
+        usleep(1000000);
     }
 
     thread_return(0); // Normal return
 
     pthread_cleanup_pop(1);
+    pthread_cleanup_pop(1);
+}
+
+void *telemetry_send_padstate(void *arg) {
+    telemetry_padstate_args_t *args = (telemetry_padstate_args_t *)arg;
+    padstate_t *state = args->state;
+
+    for (;;) {
+        struct timespec time;
+        clock_gettime(CLOCK_MONOTONIC, &time);
+        uint32_t time_ms = time.tv_sec * 1000 + time.tv_nsec / 1000000;
+
+        // sending arming update
+        header_p hdr = {.type = TYPE_TELEM, .subtype = TELEM_ARM};
+        arm_lvl_e arm_lvl;
+        padstate_get_level(state, &arm_lvl);
+        arm_state_p body = {.time = time_ms, .state = arm_lvl};
+
+        struct iovec pkt[2] = {
+            {.iov_base = &hdr, .iov_len = sizeof(hdr)},
+            {.iov_base = &body, .iov_len = sizeof(body)},
+        };
+        struct msghdr msg = {
+            .msg_iov = pkt,
+            .msg_iovlen = (sizeof(pkt) / sizeof(struct iovec)),
+        };
+        telemetry_publish(args->sock, &msg);
+
+        // sending actuator update
+        for (int i = 0; i < NUM_ACTUATORS; i++) {
+            header_p hdr = {.type = TYPE_TELEM, .subtype = TELEM_ACT};
+            bool act_state;
+            padstate_get_actstate(state, i, &act_state);
+            act_state_p body = {.time = time_ms, .id = i, .state = act_state};
+
+            struct iovec pkt[2] = {
+                {.iov_base = &hdr, .iov_len = sizeof(hdr)},
+                {.iov_base = &body, .iov_len = sizeof(body)},
+            };
+            struct msghdr msg = {
+                .msg_iov = pkt,
+                .msg_iovlen = (sizeof(pkt) / sizeof(struct iovec)),
+            };
+            telemetry_publish(args->sock, &msg);
+        }
+
+        struct timespec cond_timeout;
+        clock_gettime(CLOCK_REALTIME, &cond_timeout);
+        cond_timeout.tv_sec += PADSTATE_UPDATE_TIMEOUT_SEC;
+
+        // waiting until cond_timedwait times out
+        pthread_mutex_lock(&state->last_update.mut);
+        while (pthread_cond_timedwait(&state->last_update.cond, &state->last_update.mut, &cond_timeout) == 0) {
```

Comment:
Here you also want to include the predicate in your condition. Normally condvars do something like:

```c
while (!predicate) {
    pthread_cond_wait(&cond, &mut);
}
```

This prevents spurious wakeups where you wake up from the condwait but the condition wasn't actually signalled. This is possible behaviour and the dumbest thing in POSIX. In your case, the while loop condition would be checking for either the predicate *or* a timeout.

Reply:
I don't get this, can you link some resources about this? Or at least maybe like the official name of this behavior

Reply:
It's called spurious wakeups, you can read about it in the manpages here: https://linux.die.net/man/3/pthread_cond_wait